File: ruby/lib/coding_interview/application/usecase/asset/get_asset_usecase.rb
module CodingInterview
  module Application
    module Usecase
      module Asset
        GetAssetUsecaseInput = Struct.new(:user_id)
        GetAssetStockOutput = Struct.new(:symbol, :amount_jpy)
        GetAssetUsecaseOutput = Struct.new(:cash_amount, :stocks)

        class GetAssetUsecaseException < StandardError; end
        class UserNotFound < GetAssetUsecaseException; end

        class GetAssetUsecase
          def initialize(account_repository)
            @account_repository = account_repository
          end

          def run(input)
            account = @account_repository.find(input.user_id)
            raise UserNotFound if account.nil?
            stocks = account.stocks.map do |e|
              GetAssetStockOutput.new(e.symbol, e.amount_jpy)
            end
            GetAssetUsecaseOutput.new(account.cash, stocks)
          end
        end
      end
    end
  end
end

File: python/src/coding_interview/application/usecase/asset/get_asset_usecase.py
from __future__ import annotations

from dataclasses import dataclass
from decimal import Decimal

from coding_interview.application.repository.account_repository import AccountRepository
from coding_interview.application.usecase.exceptions import UserNotFoundError
from coding_interview.domain.stock_symbol import StockSymbol
from coding_interview.domain.user_id import UserId


@dataclass(frozen=True)
class GetAssetUsecaseInput:
    user_id: UserId


@dataclass(frozen=True)
class GetAssetStockOutput:
    symbol: StockSymbol
    amount_jpy: Decimal


@dataclass(frozen=True)
class GetAssetUsecaseOutput:
    cash_amount: Decimal
    stocks: tuple[GetAssetStockOutput, ...]


class GetAssetUsecase:
    def __init__(self, account_repository: AccountRepository) -> None:
        self._account_repository = account_repository

    def run(self, input: GetAssetUsecaseInput) -> GetAssetUsecaseOutput:
        account = self._account_repository.find(input.user_id)
        if account is None:
            raise UserNotFoundError()
        stocks = tuple(
            GetAssetStockOutput(symbol=s.symbol, amount_jpy=s.amount_jpy)
            for s in account.stocks
        )
        return GetAssetUsecaseOutput(cash_amount=account.cash, stocks=stocks)
